### ISSUE TITLE
closure upvars disable `unused_assignments` lint

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#134413]
```rust
fn main() {
    let mut overwritten = false;
    let _ = || overwritten = true;
}
```
does not trigger any lints while
```rust
fn main() {
    let mut overwritten = false;
    if false {
        overwritten = true;
    }
}
```
results in
```
warning: variable `overwritten` is assigned to, but never used
 --> src/main.rs:2:13
  |
2 |     let mut overwritten = false;
  |             ^^^^^^^^^^^
  |
  = note: consider using `_overwritten` instead
  = note: `#[warn(unused_variables)]` on by default
```

The `unused_variables` and lint should look into nested closures here.
